R/predict.R

```diff
@@ -0,0 +1,104 @@
+#' Predict hazard, cumulative hazard or survival probability
+#'
+#' @param object An object of class \code{pam_xgb}
+#' @param newdata A data set containing the same covariates as used for model
+#' fitting. If of class \code{data.frame}, the function will transform the
+#' data to the PED format using the same settings as for the data used in
+#' estimation.
+#' @param type The type of prediction desired. Either hazard (\code{type = "hazard"}),
+#' cumulative hazard (\code{type = "cumu_hazard"}) or survival probability
+#' (\code{type = "surv_prob"}).
+#' @param ... Currently not used.
+#' @importFrom stats predict
+#' @return A matrix of predictions containing one row per
+#' observation (row in newdata) and 1 column per  specified time in the
+#' \code{times} argument.
+#' @seealso pamm
+#' @export
+predict.pamm <- function(
+  object,
+  newdata,
+  type = c("hazard", "cumu_hazard", "surv_prob"),
+  ...) {
+
+  type <- match.arg(type)
+
+  if (!is.ped(newdata)) {
+    newdata <- as_ped(object, newdata)
+  }
+
+  newdata[["pred"]] <- predict(unpam(object), newdata, type = "response")
+  if (type == "cumu_hazard") {
+    newdata <- newdata %>%
+      group_by(.data$id) %>%
+      mutate(pred = cumsum(.data$pred * exp(.data$offset)))#TODO: is it correct to use offset here?
+  }
+  if (type == "surv_prob") {
+     newdata <- newdata %>%
+      group_by(.data[["id"]]) %>%
+      mutate(pred = exp(-cumsum(.data$pred * exp(.data$offset))))
+  }
+
+  newdata %>%
+    group_by(.data[["id"]]) %>%
+    filter(row_number() == n()) %>%
```

Comment:
1. why not return the entire dataframe? are there any downstream uses that rely on this being a dataframe with  as many rows as observational units instead of as many as "intervals spent under risk"? The final value is pretty useless I think (unless type = "surv_prob", ok....)
2.  suggested change (but see above re `"pred"`):
```r
if (type != "hazard") {
  newdata <- newdata %>%
    group_by(.data$id) %>%
    mutate(pred = cumsum(.data[["pred"]] * exp(.data[["offset"]])))
}
if (type == "surv_prob") {
  newdata <- newdata %>%
    mutate(pred = exp(-.data[["pred"]]))
}

newdata %>%
  pull(.data[["pred"]]) 
```

Reply:
Yeah, I just wanted to provide some S3 method for prediction for consistency. So decided to return a vector of the same length number of rows in input data (as is the case for most predict methods, e.g. lm, glm, gam). Thought about returning data frame with extra column, but this would replicate the functionality of `add_*` functions. 
Maybe return the E(survival time|covariates)? :man_shrugging:

Reply:
ah ok, sorry, obvs you want a vector return not a data.frame....   
expected survival time makes a lot of sense for `type = "response"`.
for the others, maybe a list of length n of (step)functions or a list of dataframes `cbind(tstart, tstop, estimate)` .

Reply:
The only Q is how to obtain the expected survival times :smile: 
1. calculate hazard (per interval) -> simulate survival times using `rpexp` -> use mean or median
2. Calculate survival prob over follow up -> use median survival time (could be NA if S(t|x)>.5 fora ll t 
3. Is it possible to calculate E(t|x) analytically, similar to parametric models?

Reply:
analytically should be possible I think? the density of the pexp isn't to terrible, after all... if not, expected survival should still be ok to compute since expectation of a positive random variable is simply the integral of S(t), which we know how to compute from the fit. practical problem there is what the upper limit of the integral should be (certainly not just maximal follow-up Tmax, but if the hazard estimate in the final interval is small and S(Tmax) is still >>0, extending that last interval for a long time will pull the expectation way up....)